pkg/bootkube/approve.go

```diff
@@ -0,0 +1,102 @@
+package bootkube
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api/v1"
+	certificates "k8s.io/kubernetes/pkg/apis/certificates/v1alpha1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/release_1_5/typed/certificates/v1alpha1"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+// ApproveKubeletCSRs indefinitely approves any CSRs submitted by Kubelet
+// instances, that are in the process of bootstrapping their TLS assets, without
+// making any kind of validation.
+func ApproveKubeletCSRs() error {
+	errCh := make(chan error)
+
+	client, err := clientset.NewForConfig(
+		&restclient.Config{Host: insecureAPIAddr},
+	)
+	if err != nil {
+		return err
+	}
+	csrClient := client.CertificatesV1alpha1Client.CertificateSigningRequests()
+
+	watchList := cache.NewListWatchFromClient(
+		client.CertificatesV1alpha1().RESTClient(),
+		"certificatesigningrequests",
+		v1.NamespaceAll,
+		fields.Everything(),
+	)
+
+	_, controller := cache.NewInformer(
+		watchList,
+		&certificates.CertificateSigningRequest{},
+		time.Second*30,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				if request, ok := obj.(*certificates.CertificateSigningRequest); ok {
+					approveCSR(csrClient, request, errCh)
+				}
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				if request, ok := obj.(*certificates.CertificateSigningRequest); ok {
+					approveCSR(csrClient, request, errCh)
+				}
+			},
+		},
+	)
+
+	go controller.Run(wait.NeverStop)
+	return <-errCh
+}
+
+func approveCSR(client v1alpha1.CertificateSigningRequestInterface, request *certificates.CertificateSigningRequest, errCh chan error) {
```

Comment:
other things can submit CSRs... if the intent of this is to approve kubelet certs, it should verify it was submitted by the kubelet bootstrap user and that the requested cert matches the expected shape for a kubelet cert

see https://github.com/kubernetes/kubernetes/tree/master/pkg/controller/certificates/groupapprove.go#L40